### DISCUSSION TITLE
Is it possible to do global transformations of Grids with nanovdb?

### DISCUSSION THREAD

**Discussion by stevenewbold** [AcademySoftwareFoundation/openvdb#1325] [Q&A]
I can see with openvdb that we can use a matrix to transform a grid, but I can't see a similar way of doing this with nanovdb.  Is this possible?

I'm trying to write something light-weight (hence nanovdb) for rendering volumes, but I need to be able to allow simple affine transformations of the whole grid to a different WorldSpace.  While I guess I can incorporate this myself as part of the rendering process, being able to scale/rotate/move a grid would make things a lot simpler.

Apologies if I'm simply missing something... pretty new to both openvdb and nanovdb.

**Comment by alessandroAM** [comment]
If I understood your question correctly, you can simply transform a grid (translate/rotate/scale) like in this example:

`    openvdb::math::Mat4d mat = openvdb::math::Mat4d::identity();
    openvdb::math::Transform::Ptr linearTransform = openvdb::math::Transform::createLinearTransform(mat);
    linearTransform->postTranslate(openvdb::Vec3d(100, 0, 0) );
    vdbGrid->setTransform(linearTransform);
`
Here I translate the grid 100 units on the X axis.

**Reply by stevenewbold** [reply]
This is exactly what I want to do, but exclusively with nanovdb and not openvdb.  I'm looking at nanovdb:map() to see if this can be used in a similar way by defining the world and inverse world matrix.

**Reply by iaomw** [reply]
> This is exactly what I want to do, but exclusively with nanovdb and not openvdb. I'm looking at nanovdb:map() to see if this can be used in a similar way by defining the world and inverse world matrix.

Hi, any update on how to transform nanovdb grid?

**Reply by apradhana** [reply]
`nanovdb::Map` is defined [here](https://github.com/AcademySoftwareFoundation/openvdb/blob/master/nanovdb/nanovdb/NanoVDB.h#L2220-L2310). In my application I use it to go from `coord` space to world space by applying `applyMapF`. Other examples can be found [here](https://github.com/AcademySoftwareFoundation/openvdb/blob/master/nanovdb/nanovdb/util/Ray.h#L179-L237).

**Reply by dchristopherfennell** [reply]
I'm trying to orient my nanovdb volume. Reading the above, it appears that openvdb has an easy way to do this by adding a transform.

How is this done using nanovdb? I'm looking at nanovdb::Map, but it's not straight forward on how to rotate orient the volume.

**Reply by stevenewbold** [reply]
In the end I transformed my rays into the grid space to get the appearance of transforming the grid in world space.  Would be much easier to have an easy way to apply a transformation to the grid itself.

**Reply by dchristopherfennell** [reply]
It doesn't make sense to me that there isn't an easy way to orient the volume like in openvdb. Transforming rays into grid space seems overly complicated, but as a last resort, I may have to do that.
